cdap-app-fabric/src/main/java/io/cdap/cdap/internal/operations/OperationLifecycleService.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright © 2023 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.operations;
+
+import com.google.inject.Inject;
+import io.cdap.cdap.internal.app.store.OperationRunDetail;
+import io.cdap.cdap.spi.data.StructuredTableContext;
+import io.cdap.cdap.spi.data.transaction.TransactionRunner;
+import io.cdap.cdap.spi.data.transaction.TransactionRunners;
+import java.io.IOException;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+
+/**
+ * Service that manages lifecycle of Operation.
+ */
+public class OperationLifecycleService {
+
+  private final TransactionRunner transactionRunner;
+
+  @Inject
+  public OperationLifecycleService(TransactionRunner transactionRunner) {
+    this.transactionRunner = transactionRunner;
+  }
+
+  //TODO(samik) add unit test
+  /**
+   * Scan operations in a namespace.
+   *
+   * @param request scan request including filters and limit
+   * @param txBatchSize batch size of transaction
+   * @param consumer {@link Consumer} to process each scanned run
+   */
+  public boolean scanOperations(ScanOperationRunsRequest request, int txBatchSize,
+      Consumer<OperationRunDetail<?>> consumer) throws OperationRunNotFoundException, IOException {
+
+    AtomicReference<ScanOperationRunsRequest> requestRef = new AtomicReference<>(request);
+    AtomicReference<String> lastKey = new AtomicReference<>(request.getScanAfter());
+    AtomicInteger currentLimit = new AtomicInteger(request.getLimit());
+
+    while (currentLimit.get() > 0) {
+      ScanOperationRunsRequest batchRequest = ScanOperationRunsRequest
+          .builder(requestRef.get())
+          .setScanAfter(lastKey.get())
+          .setLimit(Math.min(txBatchSize, currentLimit.get()))
+          .build();
+
+      requestRef.set(batchRequest);
+
+      TransactionRunners.run(transactionRunner, context -> {
+        lastKey.set(
```

Comment:
you can use the run method with a return value instead of setting lastKey in the closure. Then it can be a normal String instead of an AtomicReference.

Reply:
done